File: Java/Problem4.java
public class Solution{
    public boolean isMatch(String s, String p){
        Boolean[][] memo = new Boolean[s.length() + 1][p.length() + 1];

        return dp(0, 0, s, p, memo);
    }

    private boolean dp(int i, int j,String s, String p, Boolean[][] memo) {
        if (memo[i][j] != null) {
            return memo[i][j];
        }

        boolean result;

        if (j == p.length()){
            result = (i == s.length());
        } else if (i == s.length()) {

            if ((p.length() - j) % 2 != 0){
                result = false;
            } else {
                result =  true;
                for (int k = j; k + 1 < p.length(); k += 2) {
                    if (p.charAt(k + 1) != '*') {
                        result = false;
                        break;
                    }
                }
            }
        } else {
            boolean firstMatch = (s.charAt(i) == p.charAt(j)  || p.charAt(j)  == '.');

            if (j + 1 < p.length() && p.charAt(j + 1) == '*') {
                result = dp(i, j + 2, s, p, memo) || (firstMatch && dp(i + 1, j, s, p, memo));
            } else {
                result = firstMatch && dp(i + 1, j +1, s, p, memo);
            }
        }
        memo[i][j] =  result;
    return result;
    } 
}

File: Python/Problem4.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        memo = [[None] * (len(p) + 1) for _ in range(len(s) + 1) ]
        def dp(i: int, j: int) -> bool:
            if memo[i][j] is not None:
                return memo[i][j]

            if j == len(p):
                result = i == len(s)
            elif i == len(s):
                if (len(p) - j) % 2 == 1:
                    result = False
                else:
                    result = all(p[k + 1] == '*' for k in range (j, len(p), 2))
            else:
                first_match = s[i] == p[j] or p[j] == '.'
                if j + 1 < len(p) and p[j + 1] == '*':
                    result = dp(i, j + 2) or (first_match and dp(i + 1, j))
                else:
                    result = first_match and dp(i + 1,j + 1)
            memo [i][j] = result
            return result
        return dp(0,  0)
        